### PR TITLE
BlockHandler: no dynamic allocation

### PR DESCRIPTION
**Pull request by tigerw**

### CODE REVIEW

File: src/BlockInfo.h [hunk 1]
```diff
@@ -47,13 +47,14 @@ class cBlockInfo
 
 	// tolua_end
 
-	inline static cBlockHandler * GetHandler      (BLOCKTYPE a_Type) { return Get(a_Type).m_Handler.get();       }
-
 	/** Creates a default BlockInfo structure, initializes all values to their defaults */
 	cBlockInfo();
 
+	/** Gets the blockhandler for the given block type. */
+	static cBlockHandler * GetHandler(BLOCKTYPE a_Type);
```
Comment (peterbell10) [inline src/BlockInfo.h:54]: Why move this out of line and out of the blockinfo array?

Comment (tigerw) [reply src/BlockInfo.h:54]: The array doesn't store pointers to handlers anymore, and the implementation that calls cBlockHandler::GetBlockHandler is in the cpp where it knows about BlockHandler.h.

Comment (peterbell10) [reply src/BlockInfo.h:54]: >The array doesn't store pointers to handlers anymore

Yes, I can see that. I'm asking why.

Comment (tigerw) [reply src/BlockInfo.h:54]: The handlers are static and cBlockHandler knows where they are and pointers can be retrieved by GetBlockHandler, why have an extra step?

File: src/Blocks/BlockHandler.cpp [hunk 2]
```diff
@@ -1300,7 +1622,7 @@ void Temporary::RegisterAllBlockHandlers(BlockTypeRegistry & aRegistry)
 
 	for (const auto & r: regs)
```
Comment (bearbin) [inline src/Blocks/BlockHandler.cpp:1623]: No commenting out! If this isn't needed, just delete the whole thing (or at least comment out the for loop as well)!

**Review by bearbin — CHANGES_REQUESTED** [review]

File: src/BlockInfo.h (hunk 1 above)
Comment (peterbell10) [reply src/BlockInfo.h:54]: A switch statement is just another form of lookup table. There is no extra step.

File: src/Blocks/BlockHandler.cpp (hunk 2 above)
Comment (tigerw) [reply src/Blocks/BlockHandler.cpp:1623]: Ah, requesting review for #4855

File: src/BlockInfo.h (hunk 1 above)
Comment (tigerw) [reply src/BlockInfo.h:54]: Sure but you have to generate the table at runtime which then takes up some memory. BlockInfo is going to be made entirely stateless so we won't have a cBlockInfoArray anymore. The current fixed std:;array<256> isn't going to play nicely with 1.13+ or warn us if we forget to handle a block type in a switch statement.

**Comment by tigerw** [comment]
merge?

**Comment by tigerw** [comment]
last call for comments xD

**Review by bearbin — APPROVED** [review]
I never liked the big table of blocktypes, this seems like an improvement, aside from the addition of an extra name for each block, which adds potential for typos and copy-paste errors. But I can't see a way around that really, so this is fine.

**Comment by madmaxoft** [comment]
What exactly is wrong with (one-time) dynamic allocation?

**Comment by tigerw** [comment]
One thing is the compiler complains about incomplete switch if we forget to handle a block, but it gets better than that! Now that every handler is enumerated explicitly I'm going to tag everything `constexpr`, to make all init at compile time, no startup cost at all!

**Pull request merged**